File: 剑指offer/easy/5_替换空格/5_替换空格.cpp
class Solution {
public:
    string replaceSpace(string s)
    {
        int count = 0;     //统计空格的个数
        int OldSize = s.size();
        for (int i = 0; i < s.size(); i++)
        {
            if (s[i] == ' ')
            {
                count++;
            }
        }
        //扩充字符串s的大小
        s.resize(s.size() + count * 2);
        int NewSize = s.size();
        //从后先前将空格替换为"%20"
        for (int i = NewSize - 1, j = OldSize - 1; j < i; i--, j--)
        {
            if (s[j] != ' ')
            {
                s[i] = s[j];
            } 
            else
            {
                s[i] = '0';
                s[i - 1] = '2';
                s[i - 2] = '%';
                i -= 2;
            }
        }
        return s;
    }
};

File: 剑指offer/easy/5_替换空格/5_替换空格.py
class Solution:
    def replaceSpace(self, s: str) -> str:
        count = 0       #统计空格的个数
        old_size = len(s)
        
        #统计空格数量
        for c in s:
            if c == ' ':
                count += 1
        
        #扩充列表大小
        res = list(s)
        res.extend([' '] * (count * 2))
        new_size = len(res)
        
        #从后向前将空格替换为"%20"
        i = new_size - 1
        j = old_size - 1
        
        while j < i:
            if s[j] != ' ':
                res[i] = s[j]
            else:
                res[i] = '0'
                res[i - 1] = '2'
                res[i - 2] = '%'
                i -= 2
            i -= 1
            j -= 1
        
        return ''.join(res)